### PR TITLE
Release Planning and Roadmapping : Roadmap Visualization & Core Export #127

### PR DESCRIPTION
**Pull request by Gowrishbyndoor**
Issue ref: https://github.com/dpaia/feature-service/issues/127

## Objective
Implement API endpoints and data aggregation logic to support roadmap visualization and roadmap export functionality.

### Roadmap API

1) GET /api/roadmap — Product roadmap with releases, features, progress metrics, health indicators

- Filters:
  - productCode (optional)
  - startDate (optional)
  - endDate (optional)
  - includeCompleted (optional, defaults to true)
  - groupBy (optional; allowed values: PRODUCT, STATUS, ASSIGNEE)
- includeCompleted default: true. If includeCompleted=false, exclude releases with status RELEASED from the response.
- Date range validation: It is not necessary to have both startDate and endDate fields, when both startDate and endDate are provided, endDate must be strictly after startDate (not equal and not before).
- Grouping: Validate the optional groupBy parameter (case-insensitive) against PRODUCT, STATUS, and ASSIGNEE, normalize valid values to upper-case in appliedFilters, apply server-side stable null-safe sorting of features by PRODUCT (code), STATUS (status name, nulls first), or ASSIGNEE (assignedTo, nulls first), and return HTTP 400 for invalid values.
- RiskLevel values: LOW, MEDIUM, HIGH.
- TimelineAdherence values: ON_SCHEDULE, DELAYED (>30 days delay from createdAt to now), CRITICAL (>180 days delay from createdAt to now).
- Favorite calculation: isFavorite is computed by checking whether the feature’s code appears in the current user’s FavoriteFeature set.
- Favorites integrity (minimal rule): compute isFavorite per-user after in-memory filtering/grouping; use the field name "isFavorite"; return only true/false (no nulls). Applies uniformly to all endpoints and export.
- Server-side filtering only: all filters must be applied in memory on fetched data, not through database query filters.
- Input normalization: treat an empty or whitespace-only productCode query parameter as absent; appliedFilters.productCode must be null when input is empty.

Response structure:
```
{
  "roadmapItems": [
    {
      "release": {
        "id": Long,
        "code": String,
        "description": String,
        "status": String,
        "releasedAt": DateTime,
        "createdBy": String,
        "createdAt": DateTime,
        "updatedBy": String,
        "updatedAt": DateTime
      },
      "progressMetrics": {
        "totalFeatures": Integer,
        "completedFeatures": Integer,
        "inProgressFeatures": Integer,
        "newFeatures": Integer,
        "onHoldFeatures": Integer,
        "completionPercentage": Double
      },
      "healthIndicators": {
        "timelineAdherence": String,
        "riskLevel": String,
        "blockedFeatures": Integer
      },
      "features": [
        {
          "id": Long,
          "code": String,
          "title": String,
          "description": String,
          "status": String,
          "releaseCode": String,
          "isFavorite": Boolean,
          "assignedTo": String,
          "createdBy": String,
          "createdAt": DateTime,
          "updatedBy": String,
          "updatedAt": DateTime
        }
      ]
    }
  ],
  "summary": {
    "totalReleases": Integer,
    "completedReleases": Integer,
    "draftReleases": Integer,
    "totalFeatures": Integer,
    "overallCompletionPercentage": Double
  },
  "appliedFilters": {
    "productCode": String | null,
    "startDate": LocalDate,
    "endDate": LocalDate,
    "includeCompleted": Boolean,
    "groupBy": String,
    "owner": String | null
  }
}
```

2) GET /api/roadmap/multi-product — Aggregated roadmap across multiple products

- Filters:
  - productCodes (mandatory)
  - startDate (optional)
  - endDate (optional)
  - includeCompleted (optional; defaults to true)
  - groupBy (optional)

Response structure:
```
{
  "products": [
    {
      "productId": Long,
      "productName": String,
      "productCode": String,
      "roadmapItems": [<roadmapItems as above>]
    }
  ],
  "summary": <summary aggregated across all products>,
  "appliedFilters": {
    "productCodes": [String],
    "startDate": LocalDate,
    "endDate": LocalDate,
    "includeCompleted": Boolean,
    "groupBy": String
  }
}
```

3) GET /api/roadmap/by-owner — Roadmap filtered by owner

- Filters:
  - owner (mandatory)
  - productCode (optional)
  - startDate (optional)
  - endDate (optional)
  - includeCompleted (optional; defaults to true)
  - groupBy (optional)
- Owner filtering: features must be filtered by assignedTo == owner (do not include createdBy matches or empty values).
- Release filtering for owner roadmap: When displaying a roadmap for a specific owner, only show releases that contain at least one feature assigned to that owner.
- Input normalization: treat empty/blank productCode as absent; appliedFilters.productCode must be null when input is empty.

Response structure:
```
{
  "owner": String,
  "roadmapItems": [<roadmapItems same as /api/roadmap api>],
  "summary": <summary same as /api/roadmap api>,
  "appliedFilters": <appliedFilters same as /api/roadmap api>
}
```

### Export API

1) GET /api/roadmap/export?format=CSV|PDF — Export roadmap in CSV/PDF

- Filters:
  - productCodes (optional)
  - startDate (optional)
  - endDate (optional)
  - includeCompleted (optional; defaults to true)
  - groupBy (optional)
- format is mandatory; allowed values are CSV and PDF (case-insensitive).
- Filename: `Roadmap_{timestamp in yyyyMMddHHmmss}.csv` or `.pdf` (Content-Disposition attachment header must include this name).
- CSV columns (exact order):
  - Product Code
  - Product Name
  - Release Code
  - Release Description
  - Release Status
  - Released At
  - Total Features
  - Completed Features
  - In Progress Features
  - New Features
  - On Hold Features
  - Completion Percentage
  - Timeline Adherence
  - Risk Level
  - Blocked Features
  - Feature Code
  - Feature Title
  - Feature Status
  - Assigned To
  - Created At
- Row emission rules: emit one row per feature. If a release has zero features, emit a single release-level row with the feature-related columns left empty so that all releases appear in the export.
- Server-side filtering only: all export filtering must be applied on the aggregated data in memory (not via database query filters).

## Implementation

- Service Layer:
  - Implement a RoadmapService for server-side aggregation, filtering, grouping (by PRODUCT, STATUS, ASSIGNEE), and metric computations.
  - Implement a ReportingService for CSV and PDF export generation per the above filename, MIME type, and row/column requirements.

## Validations

- Any invalid input must result in HTTP 400 Bad Request.

## Test Coverage
- Integration tests for all roadmap API endpoints.
- Integration tests for roadmap export endpoint.
- Tests for filter combinations and edge cases, including empty productCode normalization to null and owner filtering using assignedTo only.
- Tests validating health metrics (timeline adherence, risk level, blocked features) and progress metrics.
- Tests validating multi-product top-level summary and appliedFilters are present and accurate.
- Tests asserting favorites behavior: isFavorite present, non-null (true/false), and consistent across endpoints and except export endpoint.

## Acceptance Criteria

- Roadmap API returns accurate, filtered, and grouped data.
- All api responses should be same as mentioned in the api response above.
- All filters mentioned in respective api are applied server-side and work correctly.
- Health and progress metrics are included and computed as specified.
- Roadmap export works for both CSV and PDF formats with exact columns and filenames.
- Export includes releases even when they have zero features (release-level rows with empty feature columns).
- By-owner endpoint filters features by assignedTo == owner (not createdBy).
- Empty productCode input is normalized to null in appliedFilters and not used for filtering.

FAIL_TO_PASS: com.sivalabs.ft.features.api.controllers.RoadmapControllerIntegrationTests

### CODE REVIEW

File: src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                createFeatureWithStatus(FeatureStatus.RELEASED));
+
+        RoadmapFilterDto filter = createBasicFilter();
+        when(releaseRepository.findAll()).thenReturn(List.of(testRelease));
+        when(featureRepository.findByReleaseCode(testRelease.getCode())).thenReturn(mostlyBlockedFeatures);
+        when(releaseMapper.toDto(testRelease)).thenReturn(testReleaseDto);
+        when(featureMapper.toDto(any(Feature.class))).thenReturn(testFeatureDtos.get(0));
+
+        // When
+        RoadmapResponseDto result = roadmapService.getRoadmap(filter);
```
Comment (innokenty) [inline src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java]: Integration tests are strongly preferred, because here we go again: 

Will an AI-agent implement the same exact Dto classes? Probably not. 

For the integration test we can fix the API definition in the task description and rely on that one, but in a unit test we need to fix all class names for that? Doesn't sound like a good idea, does it?

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,851 @@
...
+    void shouldExportCsvWithCorrectContentType() {
+        var result = mvc.get().uri("/api/roadmap/export?format=CSV").exchange();
+
+        assertThat(result).hasStatusOk().satisfies(response -> {
+            var contentType = response.getMvcResult().getResponse().getContentType();
+            assertThat(contentType).contains("text/csv");
+
+            var contentDisposition = response.getMvcResult().getResponse().getHeader("Content-Disposition");
+            assertThat(contentDisposition).contains("attachment");
+            assertThat(contentDisposition).contains("roadmap_");
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: Updated the task description with filename format

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,851 @@
...
+    @Test
+    @WithMockOAuth2User(username = "user")
+    @Sql("/test-data.sql")
+    void shouldGetRoadmapWithoutFilters() {
+        var result = mvc.get().uri("/api/roadmap").exchange();
+
+        assertThat(result)
+                .hasStatusOk()
+                .bodyJson()
+                .convertTo(RoadmapResponseDto.class)
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: Removed the Dto dependencies and chaned to pure integrations tests, this will remove the implementation dependencies for the test files. Hope with this should be resolve the mentioned issue

File: src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    private ProductRepository productRepository;
+
+    @Mock
+    private ReleaseMapper releaseMapper;
+
+    @Mock
+    private FeatureMapper featureMapper;
+
+    @InjectMocks
+    private RoadmapService roadmapService;
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java]: Removed the service unit tests and covered in integrations tests

File: src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java (hunk 1 above)
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/domain/RoadmapServiceTest.java]: Removed the service unit tests and covered in integrations tests

**Comment by Gowrishbyndoor** [comment]
> Suggested fixes:
> 
>     1. Define exact API response structure for each endpoint in the task description.
> 
>     2. Define needed file parameters for export endpoints.
> 
>     3. Convert unit tests into integration tests to remove dependency on exact Dto class definitions.

All 3 mentioned issues are fixed. Please re-review

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,835 @@
...
+            assertThat(summary.get("totalReleases")).isEqualTo(6);
+            assertThat(summary.get("completedReleases")).isEqualTo(6);
+            assertThat(summary.get("draftReleases")).isEqualTo(0);
+            assertThat(summary.get("totalFeatures")).isEqualTo(2); // Only 2 features have release_id set
+            assertThat((Double) summary.get("overallCompletionPercentage")).isBetween(0.0, 100.0);
+            assertThat((Integer) summary.get("completedReleases") + (Integer) summary.get("draftReleases"))
+                    .isEqualTo((Integer) summary.get("totalReleases"));
+
+            // Check applied filters are null/default for no filter request
+            Map<String, Object> appliedFilters = (Map<String, Object>) response.get("appliedFilters");
```
Comment (innokenty) [inline src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: Task description only mentions `roadmapItems`, but not `summary` or `appliedFilters`.

Also I find the format of the jsons that you wrote there in task description really weird, can you please remove this part below and leave only the actual response body json?
```
{ "status": "SUCCESS", "timestamp": DateTime, "data":
```

**Review by innokenty — CHANGES_REQUESTED** [review]
Looks good in general, but some response fields are still not specified. Please add specification of those, and also remove all the unneeded parts from task description (or adjust in other ways):

- extra json parts that don't actually specify response body
- extra csv columns that we are not checking (or add checks instead)
- extra information about file name like the timestamp

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
+        var result = mvc.get().uri("/api/roadmap/export?format=PDF").exchange();
+
+        assertThat(result).hasStatusOk().satisfies(response -> {
+            var contentType = response.getMvcResult().getResponse().getContentType();
+            assertThat(contentType).contains("application/pdf");
+
+            var contentDisposition = response.getMvcResult().getResponse().getHeader("Content-Disposition");
+            assertThat(contentDisposition).contains("attachment");
+            assertThat(contentDisposition).contains("Roadmap_");
+            assertThat(contentDisposition).contains(".pdf");
```
[the remark this answers is not in the document]
Comment (Gowrishbyndoor) [reply src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: Timestamp will be dynamic values so cannot assert the values. so not checking the timestamp.

But we need the timestamp in the export file name to avoid overriding the file in the second time export.

File: src/main/java/com/sivalabs/ft/features/domain/RoadmapService.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        }
+
+        // For draft releases, compare current date with created date + some threshold
+        long daysSinceCreation =
+                java.time.Duration.between(release.getCreatedAt(), now).toDays();
+        if (daysSinceCreation > 180) { // 6 months threshold - critical delay
+            return TimelineAdherence.CRITICAL;
+        } else if (daysSinceCreation > 90) { // 3 months threshold - delayed
+            return TimelineAdherence.DELAYED;
+        }
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/RoadmapService.java]: Should be 30 days here according to task description. no?

File: src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java (hunk 5 above)
Comment (innokenty) [reply src/test/java/com/sivalabs/ft/features/api/controllers/RoadmapControllerIntegrationTests.java]: Here's what I'm talking about: obviously, AI-agent didn't create those fields in the response:

```
[ERROR] Errors:
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapByOwner:177->lambda$shouldGetRoadmapByOwner$0:180 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapByOwnerWithProductFilter:750->lambda$shouldGetRoadmapByOwnerWithProductFilter$0:752 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapWithGroupBy:134->lambda$shouldGetRoadmapWithGroupBy$0:139 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapWithIncludeCompleted:112->lambda$shouldGetRoadmapWithIncludeCompleted$0:118 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapWithProductFilter:56->lambda$shouldGetRoadmapWithProductFilter$0:62 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldGetRoadmapWithoutFilters:23->lambda$shouldGetRoadmapWithoutFilters$0:30 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleAllGroupByOptions:482->lambda$shouldHandleAllGroupByOptions$0:487 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleEmptyProductCode:334->lambda$shouldHandleEmptyProductCode$0:340 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleMultipleProductCodesInSingleRequest:817->lambda$shouldHandleMultipleProductCodesInSingleRequest$0:819 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleNonExistentOwner:376->lambda$shouldHandleNonExistentOwner$0:382 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleNonExistentProductCode:353->lambda$shouldHandleNonExistentProductCode$0:359 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldHandleSpecificProductWithFeatures:519->lambda$shouldHandleSpecificProductWithFeatures$0:524 NullPointer Cannot invoke "java.util.Map.get(Object)" because "appliedFilters" is null
[ERROR]   RoadmapControllerIntegrationTests.shouldReturnConsistentSummaryAcrossFilters:832->lambda$shouldReturnConsistentSummaryAcrossFilters$0:837 NullPointer Cannot invoke "java.util.Map.get(Object)" because "summary" is null
[INFO]
[ERROR] Tests run: 77, Failures: 23, Errors: 13, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

Here is my branch with auto implementation created with your latest task description + your latest golden test:
https://github.com/toloka-partners/feature-service/tree/toloka-agent/127-roadmap-visualization-core-export-golden-test

File: src/main/java/com/sivalabs/ft/features/domain/ReportingService.java [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        }
+    }
+
+    private String[] createCsvRow(
+            ReleaseDto release, ProgressMetricsDto progress, HealthIndicatorsDto health, FeatureDto feature) {
+        List<String> row = new ArrayList<>();
+
+        // Release information
+        row.add(release.code().split("-")[0]); // Product code from release code
+        row.add(""); // Product name - would need to be passed separately
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/ReportingService.java]: What's the point of this? Why did you add `Product Name` column, if it's always empty?

**Review by innokenty — COMMENTED** [review]
Looks good! 🎉 
Left two minor comments, please answer those, but basically I would say the rest is finished.

**Review by innokenty — CHANGES_REQUESTED** [review]
this 👆 

looks like this test (and probably others) don't match the specification

**Comment by Gowrishbyndoor** [comment]
>  Have you run your new tests against agent's implementation with new task description?

I was doing it in local, so have not asked for re-review this time. 

I also got similar issue in local, i will analyze and see what is the root cause then i will ask for a review. Thanks

**Comment by innokenty** [comment]
I added a small change 👆 – not sure what's the reason to verify CSV headers order, when we can easily verify them without order too.

(added as a separate commit for visibility, we will squash it later)

**Comment by innokenty** [comment]
So the code looks good in general now, but I realized another major thing:

Why do we have 3 roadmap endpoints in the first place? They all do basically the same stuff and use basically the same code! Wouldn't it make sense to just combine them into one? This way we will just have:

* Single `GET /api/roadmap` endpoint that just does all the functionality in one place and accepts various parameters including:
  * `productCodes` - which can be one like in the first, or can be multiple like in the second
  * `owner` – which just adds another filter (and also can be simply renamed to `assignedTo`)
  
  What do you think? I think that will allow us to throw away a lot of code, and simplify task description by a lot as well.
